pil-analyzer/src/lib.rs

```diff
@@ -29,27 +29,32 @@ pub use pil_analyzer::{analyze_ast, analyze_file, analyze_string};
 
 pub trait AnalysisDriver: Clone + Copy {
     /// Turns a declaration into an absolute name.
-    fn resolve_decl(&self, name: &String) -> String {
+    fn resolve_decl(&self, name: &String) -> Result<String, String> {
         self.resolve_namespaced_decl(&[name])
-            .relative_to(&Default::default())
-            .to_string()
+            .map(|path| path.relative_to(&Default::default()).to_string())
     }
     /// Turns a nested declaration into an absolute name.
-    fn resolve_namespaced_decl(&self, path: &[&String]) -> AbsoluteSymbolPath;
-    fn resolve_value_ref(&self, path: &SymbolPath) -> String {
+    fn resolve_namespaced_decl(&self, path: &[&String]) -> Result<AbsoluteSymbolPath, String>;
+    fn resolve_value_ref(&self, path: &SymbolPath) -> Result<String, String> {
         self.resolve_ref(path, SymbolCategory::Value)
     }
-    fn resolve_type_ref(&self, path: &SymbolPath) -> String {
+    fn resolve_type_ref(&self, path: &SymbolPath) -> Result<String, String> {
         self.resolve_ref(path, SymbolCategory::Type)
     }
-    fn resolve_ref(&self, path: &SymbolPath, symbol_category: SymbolCategory) -> String {
+    fn resolve_ref(
+        &self,
+        path: &SymbolPath,
+        symbol_category: SymbolCategory,
+    ) -> Result<String, String> {
         let (path, cat) = self
             .try_resolve_ref(path)
-            .unwrap_or_else(|| panic!("{symbol_category} symbol not found: {path}"));
+            .ok_or_else(|| format!("{symbol_category} symbol not found: {path}"))?;
         if !cat.compatible_with_request(symbol_category) {
```

Comment:
Sorry if this is pedantic, but I would apply two suggestions for readability here: (1) Avoid negation in conditions if possible (turn around the cases instead) and (2) only use explicit `return` if an early return makes sense. So this would become something like
```rust
if cat.compatible... {
  Ok(path)
} else {
  Err(format!(..))
}
```